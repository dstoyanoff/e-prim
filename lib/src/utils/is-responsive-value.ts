import { ResponsiveValue } from "../system/types";
import { TBreakpoint } from "../theme";

/**
 * Check if the value is an object that contains either breakpoint
 */
export const isResponsiveValue = <T>(value: ResponsiveValue<T>, breakpoints: TBreakpoint) => {
  if (typeof value !== "object") {
    return false;
  }

  return Object.keys(breakpoints).some(breakpoint => value && breakpoint in value);
};
