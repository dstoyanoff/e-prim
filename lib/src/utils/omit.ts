export const omit = <T extends object>(obj: T, ...keys: (keyof T)[]) => {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    if (!keys.includes(key as keyof T)) {
      acc[key] = value;
    }

    return acc;
  }, {} as Record<string, unknown>) as T;
};
