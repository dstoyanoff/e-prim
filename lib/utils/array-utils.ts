/**
 * Takes either a single element or an array and always returns an array of one or many elements
 * @param items input items of generic type
 */
export const singleOrMulti = <T>(items: T | T[]): T[] => (Array.isArray(items) ? items : [items]);

type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
