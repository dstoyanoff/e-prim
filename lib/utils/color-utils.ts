/**
 * Converts a decimal percent to hex
 * @param p percent to convert. Should be an integer between 0 and 100, other values are bounded to that range.
 * Decimal values are rounded to the nearest integer
 */
export const percentToHex = (p: number) => {
  const percent = Math.max(0, Math.min(100, p)); // bound percent from 0 to 100
  const intValue = Math.round((percent / 100) * 255); // map percent to nearest integer (0 - 255)
  const hexValue = intValue.toString(16); // get hexadecimal representation

  return hexValue.padStart(2, "0"); // format with leading 0
};
