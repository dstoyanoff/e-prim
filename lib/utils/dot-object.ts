export type PathsToStringProps<T, TStopType = string> = T extends TStopType
  ? []
  : {
      [K in Extract<keyof T, string | number>]: [
        K,
        ...PathsToStringProps<T[K], TStopType>
      ];
    }[Extract<keyof T, string | number>];

export type Join<
  T extends (string | number)[],
  D extends string = "."
> = T extends []
  ? never
  : T extends [infer F]
  ? F
  : T extends [infer F, ...infer R]
  ? F extends string | number
    ? `${F}${D}${Join<Extract<R, (string | number)[]>, D>}`
    : never
  : string | number;

/**
 * Given a color key (dot-based deep path), resolves the actual color value
 * @param key dot-based path
 * @param object reference to the main object that should be used to find the value
 */
export const getValueFromKey = <
  TValue,
  TKey extends string = string,
  TObject = {}
>(
  key: TKey | undefined,
  object: TObject
): TValue | undefined => {
  if (!key) {
    return undefined;
  }

  const paths = key.split(".");

  return paths.reduce((accumulator, path) => {
    return accumulator ? accumulator[path] : (object as any)[path];
  }, undefined);
};
