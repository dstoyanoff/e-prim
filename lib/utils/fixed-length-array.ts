type ArrayLengthMutationKeys = "splice" | "push" | "pop" | "shift" | "unshift";

export type FixedLengthArray<T, L extends number, TObj = [T, ...Array<T>]> = Pick<
  TObj,
  Exclude<keyof TObj, ArrayLengthMutationKeys>
> & {
  readonly length: L;
  [I: number]: T;
  [Symbol.iterator]: () => IterableIterator<T>;
};
