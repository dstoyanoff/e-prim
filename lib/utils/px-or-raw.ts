export const pxOrRaw = (value: string | number, numberTransform: (value: number) => number = v => v) =>
  typeof value === "number" ? `${numberTransform(value)}px` : value;
