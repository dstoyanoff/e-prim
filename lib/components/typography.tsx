import type { ElementType, ReactElement } from "react";
import { combineResponsiveValues } from "@/system/combine-responsive-values";
import { type TypographySystem, createTypography, typographyPropKeys } from "@/system/typography-system";
import type { BaseTheme } from "@/theme";
import { omit } from "@/utils/omit";
import { Box, type BoxProps } from "./box";

const DEFAULT_TAG = "span" as const;

type OwnProps = TypographySystem;

export type TypographyProps<E extends ElementType> = Omit<BoxProps<E>, "typography"> & OwnProps;

export function Typography<E extends ElementType = typeof DEFAULT_TAG>(props: TypographyProps<E>): ReactElement | null {
  return (
    <Box
      label="typography"
      {...(omit(props, ...typographyPropKeys) as BoxProps<E>)}
      css={(theme: BaseTheme) => combineResponsiveValues(...createTypography(props, theme))}
    />
  );
}
