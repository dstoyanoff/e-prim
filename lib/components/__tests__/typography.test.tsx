/**
 * @jest-environment jsdom
 */

import { render } from "@testing-library/react";
import { wrapper } from "@/utils/mock-theme";
import { Typography } from "../typography";

describe("typography", () => {
  test("should render typography", () => {
    expect(
      render(
        <Typography
          variant="title.1"
          color="neutral.4"
          background="primary.normal"
          p={3}
          align="center"
          noWrap
          textOverflow="ellipsis"
        >
          child
        </Typography>,
        {
          wrapper,
        },
      ).container,
    ).toMatchSnapshot();
  });
});
