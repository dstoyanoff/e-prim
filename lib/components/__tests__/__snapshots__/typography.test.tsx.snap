// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typography should render typography 1`] = `
.emotion-0 {
  color: #898896;
  background: #00659e;
  padding: 12px;
  text-align: center;
  white-space: nowrap;
  text-overflow: ellipsis;
  font-family: Mulish;
  font-size: 24px;
  font-weight: 700;
}

<div>
  <div
    class="emotion-0"
  >
    child
  </div>
</div>
`;
